File: Freedom/微信/发现/WXShakeViewController.m

//  TLShakeViewController.m
//  Freedom
// Created by Super
#import "WXShakeViewController.h"
#import "WXShakeSettingViewController.h"
#import <XCategory/NSFileManager+expanded.h>
#define     SHAKE_HEIGHT    90
typedef NS_ENUM(NSUInteger, TLShakeButtonType) {
    TLShakeButtonTypePeople,
    TLShakeButtonTypeSong,
    TLShakeButtonTypeTV,
};
@interface WXShakeButton : UIButton
@property (nonatomic, strong) NSString *title;
@property (nonatomic, strong) NSString *iconPath;
@property (nonatomic, strong) NSString *iconHLPath;
@property (nonatomic, assign) TLShakeButtonType type;
@property (nonatomic, assign) NSUInteger msgNumber;
- (id) initWithType:(TLShakeButtonType)type title:(NSString *)title iconPath:(NSString *)iconPath iconHLPath:(NSString *)iconHLPath;
@end
@interface WXShakeButton ()
@property (nonatomic, strong) UIImageView *iconImageView;
@property (nonatomic, strong) UILabel *textLabel;
@end
@implementation WXShakeButton
- (id) initWithType:(TLShakeButtonType)type title:(NSString *)title iconPath:(NSString *)iconPath iconHLPath:(NSString *)iconHLPath{
    if (self = [super init]) {
        [self addSubview:self.iconImageView];
        [self addSubview:self.textLabel];
        [self p_addMasonry];
        self.type = type;
        self.title = title;
        self.iconPath = iconPath;
        self.iconHLPath = iconHLPath;
    }
    return self;
}
- (void)setTitle:(NSString *)title{
    _title = title;
    [self.textLabel setText:title];
}
- (void)setIconPath:(NSString *)iconPath{
    _iconPath = iconPath;
    [self.iconImageView setImage:[UIImage imageNamed:iconPath]];
}
- (void)setIconHLPath:(NSString *)iconHLPath{
    _iconHLPath = iconHLPath;
    [self.iconImageView setHighlightedImage:[UIImage imageNamed:iconHLPath]];
}
- (void)setSelected:(BOOL)selected{
    [super setSelected:selected];
    [self.iconImageView setHighlighted:selected];
    [self.textLabel setHighlighted:selected];
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.iconImageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.right.mas_equalTo(self);
        make.bottom.mas_equalTo(self.textLabel.mas_top).mas_offset(-8);
    }];
    [self.textLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.right.mas_equalTo(self);
        make.bottom.mas_equalTo(self);
    }];
    
    [self mas_updateConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self.iconImageView.mas_top);
    }];
}
#pragma mark - Getter -
- (UIImageView *)iconImageView{
    if (_iconImageView == nil) {
        _iconImageView = [[UIImageView alloc] init];
    }
    return _iconImageView;
}
- (UILabel *)textLabel{
    if (_textLabel == nil) {
        _textLabel = [[UILabel alloc] init];
        [_textLabel setFont:[UIFont systemFontOfSize:12.0f]];
        [_textLabel setTextAlignment:NSTextAlignmentCenter];
        [_textLabel setTextColor:[UIColor whiteColor]];
        [_textLabel setHighlightedTextColor:colorGreenDefault];
    }
    return _textLabel;
}
@end
@interface WXShakeViewController ()
@property (nonatomic, assign) TLShakeButtonType curType;
@property (nonatomic, strong) UIImageView *topLogoView;
@property (nonatomic, strong) UIImageView *bottomLogoView;
@property (nonatomic, strong) UIImageView *centerLogoView;
@property (nonatomic, strong) UIImageView *topLineView;
@property (nonatomic, strong) UIImageView *bottomLineView;
@property (nonatomic, strong) WXShakeButton *peopleButton;
@property (nonatomic, strong) WXShakeButton *songButton;
@property (nonatomic, strong) WXShakeButton *tvButton;
@end
@implementation WXShakeViewController
- (void)viewDidLoad{
    [super viewDidLoad];
    [self.navigationItem setTitle:@"摇一摇"];
    [self.view setBackgroundColor:RGBACOLOR(46.0, 49.0, 50.0, 1.0)];
    
    UIBarButtonItem *rightBarButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"nav_setting"] style:UIBarButtonItemStylePlain target:self action:@selector(rightBarButtonDown:)];
    [self.navigationItem setRightBarButtonItem:rightBarButton];
    
    [self.view addSubview:self.centerLogoView];
    [self.view addSubview:self.topLogoView];
    [self.view addSubview:self.bottomLogoView];
    [self.view addSubview:self.topLineView];
    [self.view addSubview:self.bottomLineView];
    [self.view addSubview:self.peopleButton];
    [self.view addSubview:self.songButton];
    [self.view addSubview:self.tvButton];
    [self p_addMasonry];
    
    self.curType = TLShakeButtonTypePeople;
}
- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    NSString *centreImageName = [[NSUserDefaults standardUserDefaults] objectForKey:@"Shake_Image_Path"];
    if (centreImageName) {
        NSString *path = [NSFileManager pathUserSettingImage:centreImageName];
        [self.centerLogoView setImage:[UIImage imageNamed:path]];
    }else{
        [self.centerLogoView setImage:[UIImage imageNamed:@"shake_logo_center"]];
    }
}
#pragma mark - Event Response
- (void)controlButtonDown:(WXShakeButton *)sender{
    if (sender.isSelected) {
        return;
    }
    self.curType = sender.type;
    [self.peopleButton setSelected:self.peopleButton.type == sender.type];
    [self.songButton setSelected:self.songButton.type == sender.type];
    [self.tvButton setSelected:self.tvButton.type == sender.type];
}
- (void)rightBarButtonDown:(UIBarButtonItem *)sender{
    WXShakeSettingViewController *shakeSettingVC = [[WXShakeSettingViewController alloc] init];
    [self setHidesBottomBarWhenPushed:YES];
    [self.navigationController pushViewController:shakeSettingVC animated:YES];
}
// 摇动手机
- (void)motionBegan:(UIEventSubtype)motion withEvent:(UIEvent *)event{
    if ([self.topLineView isHidden]) {
        [self.topLineView setHidden:NO];
        [self.bottomLineView setHidden:NO];
        [UIView animateWithDuration:0.5 animations:^{
            [self.topLogoView mas_updateConstraints:^(MASConstraintMaker *make) {
                make.bottom.mas_equalTo(self.view.mas_centerY).mas_offset(-10 - SHAKE_HEIGHT);
            }];
            [self.bottomLogoView mas_updateConstraints:^(MASConstraintMaker *make) {
                make.top.mas_equalTo(self.view.mas_centerY).mas_offset(-10 + SHAKE_HEIGHT);
            }];
            [self.view layoutIfNeeded];
        } completion:^(BOOL finished) {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.15 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                [UIView animateWithDuration:0.5 animations:^{
                    [self.topLogoView mas_updateConstraints:^(MASConstraintMaker *make) {
                        make.bottom.mas_equalTo(self.view.mas_centerY).mas_offset(-10);
                    }];
                    [self.bottomLogoView mas_updateConstraints:^(MASConstraintMaker *make) {
                        make.top.mas_equalTo(self.view.mas_centerY).mas_offset(-10);
                    }];
                    [self.view layoutIfNeeded];
                } completion:^(BOOL finished) {
                    [self.topLineView setHidden:YES];
                    [self.bottomLineView setHidden:YES];
                }];
            });
        }];
    }
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.centerLogoView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.mas_equalTo(self.view);
        make.width.mas_equalTo(self.view);
        make.height.mas_equalTo(200);
    }];
    [self.topLogoView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.mas_equalTo(self.view.mas_centerY).mas_offset(-10);
        make.centerX.mas_equalTo(self.view);
        make.width.mas_equalTo(self.view);
        make.height.mas_equalTo(150);
    }];
    [self.bottomLogoView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self.view.mas_centerY).mas_offset(-10);
        make.centerX.mas_equalTo(self.view);
        make.width.mas_equalTo(self.view);
        make.height.mas_equalTo(150);
    }];
    [self.topLineView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.mas_equalTo(self.view);
        make.top.mas_equalTo(self.topLogoView.mas_bottom);
    }];
    [self.bottomLineView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.mas_equalTo(self.view);
        make.bottom.mas_equalTo(self.bottomLogoView.mas_top);
    }];
    
    // bottom
    CGFloat widthButton = 40;
    CGFloat space = (APPW - widthButton * 3) / 4 * 1.2;
    [self.songButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.mas_equalTo(self.view);
        make.bottom.mas_equalTo(self.view).mas_offset(-15);
    }];
    [self.peopleButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(self.songButton);
        make.right.mas_equalTo(self.songButton.mas_left).mas_offset(-space);
    }];
    [self.tvButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(self.songButton);
        make.left.mas_equalTo(self.songButton.mas_right).mas_offset(space);
    }];
}
#pragma mark - Getter -
- (UIImageView *)topLogoView{
    if (_topLogoView == nil) {
        _topLogoView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"shake_logo_top"]];
        [_topLogoView setBackgroundColor:RGBACOLOR(46.0, 49.0, 50.0, 1.0)];
        [_topLogoView setContentMode:UIViewContentModeBottom];
    }
    return _topLogoView;
}
- (UIImageView *)bottomLogoView{
    if (_bottomLogoView == nil) {
        _bottomLogoView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"shake_logo_bottom"]];
        [_bottomLogoView setBackgroundColor:RGBACOLOR(46.0, 49.0, 50.0, 1.0)];
        [_bottomLogoView setContentMode:UIViewContentModeTop];
    }
    return _bottomLogoView;
}
- (UIImageView *)centerLogoView{
    if (_centerLogoView == nil) {
        _centerLogoView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"shake_logo_center"]];
        [_centerLogoView setContentMode:UIViewContentModeScaleAspectFill];
        [_centerLogoView setClipsToBounds:YES];
    }
    return _centerLogoView;
}
- (UIImageView *)topLineView{
    if (_topLineView == nil) {
        _topLineView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"shake_line_top"]];
        [_topLineView setHidden:YES];
    }
    return _topLineView;
}
- (UIImageView *)bottomLineView{
    if (_bottomLineView == nil) {
        _bottomLineView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"shake_line_bottom"]];
        [_bottomLineView setHidden:YES];
    }
    return _bottomLineView;
}
- (WXShakeButton *)peopleButton{
    if (_peopleButton == nil) {
        _peopleButton = [[WXShakeButton alloc] initWithType:TLShakeButtonTypePeople title:@"人" iconPath:@"shake_button_people" iconHLPath:@"shake_button_peopleHL"];
        [_peopleButton setSelected:YES];
        [_peopleButton addTarget:self action:@selector(controlButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _peopleButton;
}
- (WXShakeButton *)songButton{
    if (_songButton == nil) {
        _songButton = [[WXShakeButton alloc] initWithType:TLShakeButtonTypeSong title:@"歌曲" iconPath:@"shake_button_music" iconHLPath:@"shake_button_musicHL"];
        [_songButton addTarget:self action:@selector(controlButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _songButton;
}
- (WXShakeButton *)tvButton{
    if (_tvButton == nil) {
        _tvButton = [[WXShakeButton alloc] initWithType:TLShakeButtonTypeTV title:@"电视" iconPath:@"shake_button_tv" iconHLPath:@"shake_button_tvHL"];
        [_tvButton addTarget:self action:@selector(controlButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _tvButton;
}
@end

File: Freedom/微信/发现/WXShakeViewController.swift
//
//  WXShakeViewController.swift
//  Freedom
import SnapKit
import Foundation
enum TLShakeButtonType : Int {
    case people
    case song
    case tv
}
class WXShakeButton: UIButton {
    var title = "" {
        didSet {
            textLabel.text = title
        }
    }
    var iconPath = "" {
        didSet {
            iconImageView.image = iconPath.image
        }
    }
    var iconHLPath = "" {
        didSet {
            iconImageView.highlightedImage = iconHLPath.image
        }
    }
    var type = TLShakeButtonType.people
    var msgNumber: Int = 0
    var iconImageView = UIImageView()
    lazy var textLabel: UILabel = {
        let textLabel = UILabel()
        textLabel.font = UIFont.small
        textLabel.textAlignment = .center
        textLabel.textColor = UIColor.whitex
        textLabel.highlightedTextColor = UIColor.greenx
        return textLabel
    }()
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    convenience init(type: TLShakeButtonType, title: String, iconPath: String, iconHLPath: String) {
        self.init(frame: CGRect.zero)
        addSubview(iconImageView)
        addSubview(textLabel)
        self.type = type
        self.title = title
        self.iconPath = iconPath
        self.iconHLPath = iconHLPath
//        p_addMasonry()
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    override var isSelected: Bool {
        didSet {
            iconImageView.isHighlighted = isSelected
            textLabel.isHighlighted = isSelected
        }
    }
//    func p_addMasonry() {
//        iconImageView.snp.makeConstraints { make in
//            make.left.and().right().equalTo(self)
//            make.bottom.equalTo(self.textLabel.mas_top).mas_offset(-8)
//        })
//        textLabel.snp.makeConstraints { make in
//            make.left.and().right().equalTo(self)
//            make.bottom.equalTo(self)
//        })
//        mas_updateConstraints({ make in
//            make.top.equalTo(self.iconImageView.mas_top)
//        })
//    }
}
class WXShakeViewController: WXBaseViewController {
    var curType = TLShakeButtonType.people
    lazy var topLogoView: UIImageView = {
        let topLogoView = UIImageView(image: WXImage.shakeTop.image)
        topLogoView.backgroundColor = UIColor.back
        topLogoView.contentMode = .bottom
        return topLogoView
    }()
    lazy var bottomLogoView: UIImageView = {
        let bottomLogoView = UIImageView(image: WXImage.shakeBottom.image)
        bottomLogoView.backgroundColor = UIColor.back
        bottomLogoView.contentMode = .top
        return bottomLogoView
    }()
    lazy var centerLogoView: UIImageView = {
        let centerLogoView = UIImageView(image: WXImage.logoCenter.image)
        centerLogoView.contentMode = .scaleAspectFill
        centerLogoView.clipsToBounds = true
        return centerLogoView
    }()
    var topLineView = UIImageView(image: WXImage.shakeTop.image)
    var bottomLineView = UIImageView(image: WXImage.shakeBottom.image)
    lazy var peopleButton: WXShakeButton = {
        let peopleButton = WXShakeButton(type: .people, title: "人", iconPath: WXImage.shakePeople.rawValue, iconHLPath:WXImage.shakePeopleHL.rawValue)
        peopleButton.isSelected = true
        peopleButton.addTarget(self, action: #selector(self.controlButtonDown(_:)), for: .touchUpInside)
        return peopleButton
    }()
    lazy var songButton: WXShakeButton = {
        let songButton = WXShakeButton(type: .song, title: "歌曲", iconPath: "shake_button_music", iconHLPath: "shake_button_musicHL")
        songButton.addTarget(self, action: #selector(self.controlButtonDown(_:)), for: .touchUpInside)
        return songButton
    }()
    lazy var tvButton: WXShakeButton = {
        let tvButton = WXShakeButton(type: .tv, title: "电视", iconPath: "shake_button_tv", iconHLPath: "shake_button_tvHL")
        tvButton.addTarget(self, action: #selector(self.controlButtonDown(_:)), for: .touchUpInside)
        return tvButton
    }()
    override func viewDidLoad() {
        super.viewDidLoad()
        topLineView.isHidden = true
        bottomLineView.isHidden = true
        navigationItem.title = "摇一摇"
        view.backgroundColor = UIColor.back
        let rightBarButton = UIBarButtonItem(image: Image.setting.image, style: .plain, target: self, action: #selector(self.rightBarButtonDown(_:)))
        navigationItem.rightBarButtonItem = rightBarButton
        view.addSubview(centerLogoView)
        view.addSubview(topLogoView)
        view.addSubview(bottomLogoView)
        view.addSubview(topLineView)
        view.addSubview(bottomLineView)
        view.addSubview(peopleButton)
        view.addSubview(songButton)
        view.addSubview(tvButton)
        curType = .people
        p_addMasonry()
    }
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        let centreImageName = UserDefaults.standard.value(forKey: "Shake_Image_Path") as? String
        if let centreImageName = centreImageName {
            let path = FileManager.pathUserSettingImage(centreImageName)
            centerLogoView.image = path?.image
        } else {
            centerLogoView.image = WXImage.logoCenter.image
        }
    }
    func controlButtonDown(_ sender: WXShakeButton) {
        if !sender.isSelected {
            return
        }
        curType = sender.type
        peopleButton.isSelected = peopleButton.type == sender.type
        songButton.isSelected = songButton.type == sender.type
        tvButton.isSelected = tvButton.type == sender.type
    }
    func motionBegan(_ motion: UIEvent.EventSubtype, with event: UIEvent) {
        if topLineView.isHidden {
            topLineView.isHidden = false
            bottomLineView.isHidden = false
            UIView.animate(withDuration: 0.5, animations: {
                self.topLogoView.snp.updateConstraints({ (make) in
                    make.bottom.equalTo(self.view.snp.centerY).offset(-10)
                })
                self.bottomLogoView.snp.updateConstraints({ (make) in
                    make.top.equalTo(self.view.snp.centerY).offset(-10 + 90)
                })
                self.view.layoutIfNeeded()
            }) { finished in
                DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + Double(Int64(0.15 * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC), execute: {
                    UIView.animate(withDuration: 0.5, animations: {
                        self.topLogoView.snp.updateConstraints({ (make) in
                            make.bottom.equalTo(self.view.snp.centerY).offset(-10)
                        })
                        self.bottomLogoView.snp.updateConstraints({ (make) in
                            make.top.equalTo(self.view.snp.centerY).offset(-10)
                        })
                        self.view.layoutIfNeeded()
                    }) { finished in
                        self.topLineView.isHidden = true
                        self.bottomLineView.isHidden = true
                    }
                })
            }
        }
    }
    func rightBarButtonDown(_ sender: UIBarButtonItem) {
        let shakeSettingVC = WXShakeSettingViewController()
        navigationController?.pushViewController(shakeSettingVC, animated: true)
    }
    func p_addMasonry() {
        centerLogoView.snp.makeConstraints { make in
            make.center.equalToSuperview()
            make.width.equalTo(self.view)
            make.height.equalTo(200)
        }
        topLogoView.snp.makeConstraints { make in
            make.bottom.equalTo(self.view.snp.centerY).offset(-10)
            make.centerX.equalTo(self.view)
            make.width.equalTo(self.view)
            make.height.equalTo(150)
        }
        bottomLogoView.snp.makeConstraints { make in
            make.top.equalTo(self.view.snp.centerY).offset(-10)
            make.centerX.equalTo(self.view)
            make.width.equalTo(self.view)
            make.height.equalTo(150)
        }
        topLineView.snp.makeConstraints { make in
            make.width.equalTo(self.view)
            make.top.equalTo(self.topLogoView.snp.bottom)
        }
        bottomLineView.snp.makeConstraints { make in
            make.width.equalTo(self.view)
            make.bottom.equalTo(self.bottomLogoView.snp.top)
        }
        let widthButton: CGFloat = 40
        let space: CGFloat = (APPW - widthButton * 3) / 4 * 1.2
        songButton.snp.makeConstraints { make in
            make.centerX.equalTo(self.view)
            make.bottom.equalTo(self.view).offset(-15)
        }
        peopleButton.snp.makeConstraints { make in
            make.centerY.equalTo(self.songButton)
            make.right.equalTo(self.songButton.snp.left).offset(-space)
        }
        tvButton.snp.makeConstraints { make in
            make.centerY.equalTo(self.songButton)
            make.left.equalTo(self.songButton.snp.right).offset(space)
        }
    }
}
